build-system/common/esbuild-babel.js

```diff
@@ -16,14 +16,75 @@
 
 const babel = require('@babel/core');
 const crypto = require('crypto');
-const fs = require('fs');
+const fs = require('fs-extra');
 const path = require('path');
 
+/**
+ * Directory where the babel filecache lives.
+ */
+const CACHE_DIR = path.resolve(__dirname, '..', '..', '.babel-cache');
+
+/**
+ * Cache for storing transformed files on both memory and on disk.
+ */
+class BabelTransformCache {
+  constructor() {
+    fs.ensureDirSync(CACHE_DIR);
+
+    /** @type {Map<string, Promise<{contents: string}>>} */
+    this.map = new Map();
+
+    /** @type {Set<string>} */
+    this.fsCache = new Set(fs.readdirSync(CACHE_DIR));
+  }
+
+  getKey_(hash) {
+    return `${hash}.json`;
+  }
+
+  /**
+   * @param {string} hash
+   * @return {null | Promise<{contents: string}>}
+   */
+  get(hash) {
+    const key = this.getKey_(hash);
+    const cached = this.map.get(key);
+    if (cached) {
+      return cached;
+    }
+    if (this.fsCache.has(key)) {
+      const transformedPromise = fs.readJson(path.join(CACHE_DIR, key));
+      this.map.set(key, transformedPromise);
+      return transformedPromise;
+    }
+
+    return null;
+  }
+
+  /**
+   * @param {string} hash
+   * @param {Promise<{contents: string}>} transformPromise
+   * @return {Promise}
+   */
+  async set(hash, transformPromise) {
+    const key = this.getKey_(hash);
+    if (this.map.has(key)) {
+      throw new Error(
+        `Read race occured. Attempting to transform a file twice.`
+      );
+    }
+
+    this.map.set(key, transformPromise);
+    const transformed = await transformPromise;
+    await fs.outputJson(path.join(CACHE_DIR, key), transformed);
```

Comment:
Nit: no need to wait for this, we can just fire and forget.
```suggestion
    fs.outputJson(path.join(CACHE_DIR, key), transformed);
```